ml-agents/mlagents/trainers/trainer_controller.py

```diff
@@ -283,61 +296,40 @@ def take_step(
-        for brain_name, trainer in self.trainers.items():
-            action_info = trainer.get_action(curr_info[brain_name])
-            take_action_vector[brain_name] = action_info.action
-            take_action_memories[brain_name] = action_info.memory
-            take_action_text[brain_name] = action_info.text
-            take_action_value[brain_name] = action_info.value
-            take_action_outputs[brain_name] = action_info.outputs
         time_start_step = time()
-        new_info = env.step(
-            vector_action=take_action_vector,
-            memory=take_action_memories,
-            text_action=take_action_text,
-            value=take_action_value,
-        )
+        new_step_infos = env.step()
         delta_time_step = time() - time_start_step
+
+        for step_info in new_step_infos:
+            for brain_name, trainer in self.trainers.items():
+                if brain_name in self.trainer_metrics:
+                    self.trainer_metrics[brain_name].add_delta_step(delta_time_step)
+                trainer.add_experiences(
+                    step_info.last_all_brain_info,
+                    step_info.current_all_brain_info,
+                    step_info.all_action_infos[brain_name].outputs,
+                )
+                trainer.process_experiences(
+                    step_info.last_all_brain_info, step_info.current_all_brain_info
+                )
         for brain_name, trainer in self.trainers.items():
             if brain_name in self.trainer_metrics:
                 self.trainer_metrics[brain_name].add_delta_step(delta_time_step)
-            trainer.add_experiences(
-                curr_info, new_info, take_action_outputs[brain_name]
-            )
-            trainer.process_experiences(curr_info, new_info)
             if (
                 trainer.is_ready_update()
                 and self.train_model
                 and trainer.get_step <= trainer.get_max_steps
             ):
+                if self.train_model and trainer.get_step <= trainer.get_max_steps:
+                    trainer.increment_step(len(new_step_infos))
                 # Perform gradient descent with experience buffer
-
                 trainer.update_policy()
-            # Write training statistics to Tensorboard.
-            delta_train_start = time() - self.training_start_time
-            if self.meta_curriculum is not None:
-                trainer.write_summary(
-                    self.global_step,
-                    delta_train_start,
-                    lesson_num=self.meta_curriculum.brains_to_curriculums[
-                        brain_name
-                    ].lesson_num,
-                )
-            else:
-                trainer.write_summary(self.global_step, delta_train_start)
-            if self.train_model and trainer.get_step <= trainer.get_max_steps:
-                trainer.increment_step()
-        return new_info
+                env.set_policy(brain_name, trainer.policy)
```

Comment:
Why do you need to set the policies here?

Reply:
For the single-machine case you don't need to.  However from the perspective of the TrainerController any modification of the policy should result in a `set_policy` on the EnvManager.  In the distributed case, `set_policy` might actually be sending the serialized policy to another machine so this is a way of triggering that update.

Another way to do this would be to have a listener on policy updates, but this seemed simpler to me for this use case.

Reply:
Sounds good. And I guess theoretically a trainer could completely change its policy between steps.